share/man/man4/pkginfo.4
'\" te
.\" Copyright (c) 2003, Sun Microsystems, Inc. All Rights Reserved.
.\" Copyright 1989 AT&T
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH PKGINFO 4 "Mar 24, 2006"
.SH NAME
pkginfo \- package characteristics file
.SH DESCRIPTION
.sp
.LP
\fBpkginfo\fR is an \fBASCII\fR file that describes the characteristics of the
package along with information that helps control the flow of installation. It
is created by the software package developer.
.sp
.LP
Each entry in the \fBpkginfo\fR file is a line that establishes the value of a
parameter in the following form:
.sp
.in +2
.nf
PARAM="\fIvalue\fR"
.fi
.in -2

.sp
.LP
There is no required order in which the parameters must be specified within the
file. The \fBPKG\fR, \fBNAME\fR, \fBARCH\fR, \fBVERSION\fR and \fBCATEGORY\fR
parameters are mandatory. Other parameters are optional.
.sp
.LP
\fBpkginfo\fR provides optional parameters and an environment variable in
support of the zones (multiple Solaris environments) feature. See
\fBzones\fR(5).
.sp
.LP
The following paramaters are mandatory:
.sp
.ne 2
.na
\fB\fBARCH\fR\fR
.ad
.sp .6
.RS 4n
A comma-separated list of alphanumeric tokens that indicate the architecture
associated with the package. The \fBpkgmk\fR(1) tool can be used to create or
modify this value when actually building the package. The maximum length of a
token is 16 characters and it cannot include a comma.
.sp
Solaris's installation software meaningfully uses only one architecture token
of the form:
.sp
.in +2
.nf
<\fIinstruction_set_architecture\fR>[.<\fIplatform_group\fR>]
.fi
.in -2

where \fIplatform_group\fR is intended only for Solaris installation packages.
Third party application software should restrict itself to \fBARCH\fR values
from the following Solaris-supported instruction set architectures (\fBuname
-p\fR): \fBsparc\fR, \fBi386\fR, and \fBppc\fR. Examples of Solaris' platform
groups (\fBuname -m\fR) are \fBsun4u\fR for the \fBSPARC\fR instruction set and
\fBi86pc\fR for the i386 instruction set. See \fBuname\fR(1) and
\fBisalist\fR(1) for more details.
.RE

.sp
.ne 2
.na
\fB\fBCATEGORY\fR\fR
.ad
.sp .6
.RS 4n
A comma-separated list of categories under which a package can be displayed. A
package must at least belong to the system or application category. Categories
are case-insensitive and can contain only alphanumerics. Each category is
limited in length to 16 characters.
.RE

.sp
.ne 2
.na
\fB\fBNAME\fR\fR
.ad
.sp .6
.RS 4n
Text that specifies the package name (maximum length of 256 \fBASCII\fR
characters). Use the \fBNAME\fR parameter as the foundation for describing the
functionality and purpose of the package; spell out any acronyms and avoid
internal product/project code names. The \fBDESC\fR parameter can then be used
to expand the descriptive information. Use the \fBNAME\fR parameter to state as
specifically as possible the use of the package, why a user would need to load
it, and so on.
.RE

.sp
.ne 2
.na
\fB\fBPKG\fR\fR
.ad
.sp .6
.RS 4n
Abbreviation for the package being installed. All characters in the
abbreviation must be alphanumeric. You can also use the \fB\(mi\fR and \fB+\fR
characters in the abbreviation. The first character cannot be numeric, a
\fB+\fR or a \fB-\fR.
.sp
The abbreviation is limited to a maximum length of 32 characters.
\fBinstall\fR, \fBnew\fR, and \fBall\fR are reserved abbreviations. It is
customary to make the first four letters unique to your company, such as the
company's stock symbol.
.RE

.sp
.ne 2
.na
\fB\fBVERSION\fR\fR
.ad
.sp .6
.RS 4n
Text that specifies the current version associated with the software package.
The maximum length is 256 \fBASCII\fR characters and the first character cannot
be a left parenthesis. The \fBpkgmk\fR(1) tool can be used to create or modify
this value when actually building the package. Current Solaris software
practice is to assign this parameter monotonically increasing Dewey decimal
values of the form:
.sp
.in +2
.nf
<\fImajor_revision\fR>.<\fIminor_revision\fR>[.<\fImicro_revision\fR>]
.fi
.in -2

where all the revision fields are integers. The versioning fields can be
extended to an arbitrary string of numbers in Dewey-decimal format, if
necessary.
.RE

.sp
.LP
The following parameters are optional:
.sp
.ne 2
.na
\fB\fBBASEDIR\fR\fR
.ad
.sp .6
.RS 4n
The pathname to a default directory where "relocatable" files can be installed.
If blank, the package is not relocatable and any files that have relative
pathnames are not installed. An administrator can override the default
directory.
.RE

.sp
.ne 2
.na
\fB\fBCLASSES\fR\fR
.ad
.sp .6
.RS 4n
A space-separated list of classes defined for a package. The order of the list
determines the order in which the classes are installed. Classes listed first
are installed first (on a media by media basis). This parameter can be modified
by the request script.
.RE

.sp
.ne 2
.na
\fB\fBDESC\fR\fR
.ad
.sp .6
.RS 4n
Text that describes the package (maximum length of 256 \fBASCII\fR characters).
This parameter value is used to provide the installer with a description of
what the package contains and should build on the description provided in the
\fBNAME\fR parameter. Try to make the two parameters work together so that a
\fBpkginfo\fR \fB-l\fR provides a fairly comprehensive textual description of
the package.
.RE

.sp
.ne 2
.na
\fB\fBEMAIL\fR\fR
.ad
.sp .6
.RS 4n
An electronic address where further information is available or bugs can be
reported (maximum length of 256 \fBASCII\fR characters).
.RE

.sp
.ne 2
.na
\fB\fBHOTLINE\fR\fR
.ad
.sp .6
.RS 4n
Phone number and/or mailing address where further information can be received
or bugs can be reported (maximum length of 256 \fBASCII\fR characters).
.RE

.sp
.ne 2
.na
\fB\fBINTONLY\fR\fR
.ad
.sp .6
.RS 4n
Indicates that the package should only be installed interactively when set to
any non-null value.
.RE

.sp
.ne 2
.na
\fB\fBISTATES\fR\fR
.ad
.sp .6
.RS 4n
A list of allowable run states for package installation (for example, "\fBS s
1\fR" allows run states of \fBS\fR, \fBs\fR or \fB1\fR). The Solaris operating
environment supports the run levels \fBs\fR, \fBS\fR, \fB0\fR, \fB1\fR,
\fB2\fR, \fB3\fR, \fB5\fR, and \fB6\fR. Applicable run levels for this
parameter are \fBs\fR, \fBS\fR, \fB1\fR, \fB2\fR, and \fB3\fR. See
\fBinit\fR(1M) for details.
.RE

.sp
.ne 2
.na
\fB\fBMAXINST\fR\fR
.ad
.sp .6
.RS 4n
The maximum number of package instances that should be allowed on a machine at
the same time. By default, only one instance of a package is allowed. This
parameter must be set in order to have multiple instances of a package. In
order to support multiple instances of packages (for example, packages that
differ in their \fBARCH\fR or \fBVERSION\fR parameter value), the value of this
parameter must be high enough to allow for all instances of a given package,
including multiple versions coexisting on a software server.
.RE

.sp
.ne 2
.na
\fB\fBORDER\fR\fR
.ad
.sp .6
.RS 4n
A list of classes defining the order in which they should be put on the medium.
Used by \fBpkgmk\fR(1) in creating the package. Classes not defined in this
field are placed on the medium using the standard ordering procedures.
.RE

.sp
.ne 2
.na
\fB\fBPSTAMP\fR\fR
.ad
.sp .6
.RS 4n
Production stamp used to mark the \fBpkgmap\fR(4) file on the output volumes.
Provides a means for distinguishing between production copies of a version if
more than one is in use at a time. If \fBPSTAMP\fR is not defined, the default
is used. The default consists of the UNIX system machine name followed by the
string "\fIYYYYMMDDHHMMSS\fR" (year, month, date, hour, minutes, seconds).
.RE

.sp
.ne 2
.na
\fB\fBRSTATES\fR\fR
.ad
.sp .6
.RS 4n
A list of allowable run states for package removal (for example, "\fBS s 1\fR"
allows run states of \fBS\fR, \fBs\fR or \fB1\fR). The Solaris operating
environment supports the run levels \fBs\fR, \fBS\fR, \fB0\fR, \fB1\fR,
\fB2\fR, \fB3\fR, \fB5\fR, and \fB6\fR. Applicable run levels for this
parameter are \fBs\fR, \fBS\fR, \fB1\fR, \fB2\fR, and \fB3\fR See
\fBinit\fR(1M) for details.
.RE

.sp
.ne 2
.na
\fB\fBSUNW_ISA\fR\fR
.ad
.sp .6
.RS 4n
Solaris-only optional parameter that indicates a software package contains
64-bit objects if it is set to \fBsparcv9\fR. If this parameter is not set, the
default \fBISA\fR (instruction set architecture) is set to the value of the
\fBARCH\fR parameter.
.RE

.sp
.ne 2
.na
\fB\fBSUNW_LOC\fR\fR
.ad
.sp .6
.RS 4n
Solaris-only optional parameter used to indicate a software package containing
localization files for a given product or application. The parameter value is a
comma-separated list of locales supported by a package. It is only used for
packages containing localization files, typically the message catalogues. The
allowable values for this string field are those found in the table of Standard
Locale Names located in the \fIInternational Language Environments Guide\fR.
.sp
.in +2
.nf
SUNW_LOC="<\fIlocale_name\fR>,<\fIlocale_name\fR>,..,\e
<\fIlocale_name\fR>"
.fi
.in -2

where
.sp
.in +2
.nf
<\fIlocale_name\fR>::= <\fIlanguage\fR>[_<\fIterritory\fR>]\e
[.<\fIcodeset\fR>]
<\fIlanguage\fR>::= the set of names from ISO 639
<\fIterritory\fR>::= the set of territories specified
in ISO 3166
<\fIcodeset\fR>::= is a string corresponding to the coded
character set
.fi
.in -2

Since a value of C specifies the traditional UNIX system behavior (American
English, en_US), packages belonging to the C locale are viewed as non-localized
packages, and thus must not have \fBSUNW_LOC\fR and \fBSUNW_PKGLIST\fR included
in their \fBpkginfo\fR file. See also the \fBSUNW_LOC\fR parameter in
\fBpackagetoc\fR(4) and \fBsetlocale\fR(3C) for more information. This keyword
is not recognized by the add-on software utility Software Manager.
.RE

.sp
.ne 2
.na
\fB\fBSUNW_PKG_DIR\fR\fR
.ad
.sp .6
.RS 4n
A value set by \fBpkgadd\fR that contains the location of the installing
package. This value is provided to any install time package procedure scripts
that need to know where the installing package is located. This parameter
should never be set manually from within a \fBpkginfo\fR file.
.RE

.sp
.ne 2
.na
\fB\fBSUNW_PKG_ALLZONES\fR\fR
.ad
.sp .6
.RS 4n
Defines whether a package, when installed, must be installed and must be
identical in all zones. Assigned value can be \fBtrue\fR or \fBfalse\fR. The
default value is \fBfalse\fR. The setting of \fBSUNW_PKG_ALLZONES\fR has the
effects described below.
.sp
If set to \fBtrue\fR, the following conditions are in effect:
.RS +4
.TP
.ie t \(bu
.el o
The package must be installed in the global zone.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The package must be installed in any non-global zone that is created.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The package must be identical in all zones.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The package can be installed only by the global zone administrator.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The package cannot be installed by a non-global zone administrator.
.RE
If set to \fBfalse\fR, the following conditions are in effect:
.RS +4
.TP
.ie t \(bu
.el o
The package is not required to be installed in all zones.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The package is not required to be identical across all zones.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The package can be installed by the global zone administrator or by a
non-global zone administrator.
.RE
Packages that must be identical across all zones must set this variable to
\fBtrue\fR. This would include packages that deliver components that are part
of the core operating system, or that are dependent on interfaces exported by
the core operating system, or that deliver device drivers, or runtime libraries
that use or export operating system interfaces that are not guaranteed to be
stable across minor releases.
.sp
Packages that deliver components that are not part of the core operating system
(such as application programs) that can be different between any two zones must
set this variable to \fBfalse\fR.
.sp
With respect to \fBSUNW_PKG_ALLZONES\fR, keep in mind the following:
.RS +4
.TP
.ie t \(bu
.el o
Use of \fBpkgadd\fR in the global zone installs packages in all zones unless
\fB-G\fR is specified, in which case packages are installed in the global zone
only. The setting of \fBSUNW_PKG_ALLZONES\fR does not change this behavior. For
example, a package that has a setting of \fBSUNW_PKG_ALLZONES=false\fR is not
installed in the global zone only.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The \fBSUNW_PKG_ALLZONES\fR attribute controls whether a package \fBmust\fR be
installed in all zones (and must be the same in all zones) when it is
installed.
.RE
.RS +4
.TP
.ie t \(bu
.el o
Use of the \fB-G\fR option to \fBpkgadd\fR with a package that has
\fBSUNW_PKG_ALLZONES=true\fR is an error and causes installation of that
package to fail.
.RE
.RE

.sp
.ne 2
.na
\fB\fBSUNW_PKG_HOLLOW\fR\fR
.ad
.sp .6
.RS 4n
Defines whether a package should be visible in any non-global zone if that
package is required to be installed and be identical in all zones (for example,
a package that has \fBSUNW_PKG_ALLZONES=true\fR). Assigned value can be
\fBtrue\fR or \fBfalse\fR. The default value is \fBfalse\fR. The package is not
required to be installed, but if it is installed, the setting of
\fBSUNW_PKG_HOLLOW\fR has the effects described below.
.sp
If set to \fBfalse\fR, the following conditions are in effect:
.RS +4
.TP
.ie t \(bu
.el o
If installed in the global zone, the package content and installation
information are required in all non-global zones.
.RE
.RS +4
.TP
.ie t \(bu
.el o
Software delivered by the package is visible in all non-global zones. An
example of such a a package is the package that delivers the \fBtruss\fR(1)
command.
.RE
If set to \fBtrue\fR, the following conditions are in effect:
.RS +4
.TP
.ie t \(bu
.el o
The package content is not delivered on any non-global zone. However, the
package installation information is required on all non-global zones.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The package delivers software that should not be visible in all non-global
zones. Examples include kernel drivers and system configuration files that work
only in the global zone. This setting allows the non-global zone to resolve
dependencies on packages that are installed only in the global zone without
actually installing the package data.
.RE
.RS +4
.TP
.ie t \(bu
.el o
In the global zone, the package is recognized as having been installed, and all
components of the package are installed. Directories are created, files are
installed, and class action and other scripts are run as appropriate when the
package is installed.
.RE
.RS +4
.TP
.ie t \(bu
.el o
In a non-global zone, the package is recognized as having been installed, but
no components of the package are installed. No directories are created, no
files are installed, and no class action or other install scripts are run when
the package is installed.
.RE
.RS +4
.TP
.ie t \(bu
.el o
When removed from the global zone, the package is recognized as having been
completely installed. Appropriate directories and files are removed, and class
action or other install scripts are run when the package is removed.
.RE
.RS +4
.TP
.ie t \(bu
.el o
When removed from a non-global zone, the package is recognized as not having
been completely installed. No directories are removed, no files are removed,
and no class action or other install scripts are run when the package is
removed.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The package is recognized as being installed in all zones for purposes of
dependency checking by other packages that rely on this package being
installed.
.RE
If \fBSUNW_PKG_ALLZONES\fR is set to \fBfalse\fR, the value of this variable
has no meaning. It is a package construction error to set
\fBSUNW_PKG_ALLZONES\fR to \fBfalse\fR, then set \fBSUNW_PKG_HOLLOW\fR to
\fBtrue\fR.
.RE

.sp
.ne 2
.na
\fB\fBSUNW_PKG_THISZONE\fR\fR
.ad
.sp .6
.RS 4n
Defines whether a package must be installed in the current zone only. Assigned
value can be \fBtrue\fR or \fBfalse\fR. The default value is \fBfalse\fR. The
setting of \fBSUNW_PKG_THISZONE\fR has the effects described below.
.sp
If set to true, the following conditions are in effect:
.RS +4
.TP
.ie t \(bu
.el o
The package is installed in the current zone only.
.RE
.RS +4
.TP
.ie t \(bu
.el o
If installed in the global zone, the package is not added to any currently
existing or yet-to-be-created non-global zones. This is the same behavior that
would occur if the \fB-G\fR option were specified to \fBpkgadd\fR.
.RE
If set to false, the following conditions are in effect:
.RS +4
.TP
.ie t \(bu
.el o
If \fBpkgadd\fR is run in a non-global zone, the package is installed in the
current zone only.
.RE
.RS +4
.TP
.ie t \(bu
.el o
If \fBpkgadd\fR is run in the global zone, the package is installed in the
global zone, and is also installed in all currently installed non-global zones.
In addition, the package will be propagated to all future, newly installed
non-global zones.
.RE
.RE

.sp
.ne 2
.na
\fB\fBSUNW_PKGLIST\fR\fR
.ad
.sp .6
.RS 4n
Solaris-only optional parameter used to associate a localization package to the
package(s) from which it is derived. It is required whenever the \fBSUNW_LOC\fR
parameter is defined. This parameter value is an comma-separated list of
package abbreviations of the form:
.sp
.in +2
.nf
SUNW_PKGLIST="\fIpkg1\fR[\fI:version\fR],\fIpkg2\fR[:\fIversion\fR],..."
.fi
.in -2

where \fIversion\fR (if specified) should match the version string in the base
package specified (see \fBVERSION\fR parameter in this manual page). When in
use, \fBSUNW_PKGLIST\fR helps determine the order of package installation. The
packages listed in the parameter are installed before the localization package
in question is installed. When left blank, \fBSUNW_PKGLIST="\|"\fR, the package
is assumed to be required for the locale to function correctly. See the
\fBSUNW_PKGLIST\fR parameter in \fBpackagetoc\fR(4) for more information. This
keyword is not recognized by the add-on software utility Software Manager.
.RE

.sp
.ne 2
.na
\fB\fBSUNW_PKGTYPE\fR\fR
.ad
.sp .6
.RS 4n
Solaris-only parameter for Sun internal use only. Required for packages part of
the Solaris operating environment releases which install into the \fB/\fR,
\fB/usr\fR, and \fB/usr/openwin\fR file systems. The Solaris
operating environment installation software must know which packages are part
of which file system to properly install a server/client configuration. The
currently allowable values for this parameter are \fBroot\fR, \fBusr\fR,
\fBkvm\fR, and \fBow\fR. If no \fBSUNW_PKGTYPE\fR parameter is present, the
package is assumed to be of \fBBASEDIR=\fR\fI/opt\fR. \fBSUNW_PKGTYPE\fR is
optional only for packages which install into the \fB/opt\fR name space as is
the case for the majority of Solaris add-on software. See the
\fBSUNW_PKGTYPE\fR parameter in \fBpackagetoc\fR(4) for further information.
.RE

.sp
.ne 2
.na
\fB\fBSUNW_PKGVERS\fR\fR
.ad
.sp .6
.RS 4n
Solaris-only parameter indicating of version of the Solaris operating
environment package interface.
.sp
.in +2
.nf
SUNW_PKGVERS="<\fIsunw_package_version\fR>"
.fi
.in -2

where <\fIunw_package_version\fR> has the form \fIx.y[.z]\fR and \fIx\fR,
\fIy\fR, and z are integers. For packages built for this release and previous
releases, use \fBSUNW_PKGVERS="\fI1.0\fR"\fR.
.RE

.sp
.ne 2
.na
\fB\fBSUNW_PRODNAME\fR\fR
.ad
.sp .6
.RS 4n
Solaris-only parameter indicating the name of the product this package is a
part of or comprises (maximum length of 256 \fBASCII\fR characters). A few
examples of currently used \fBSUNW_PRODNAME\fR values are: \fB"SunOS"\fR,
\fB"OpenWindows"\fR, and \fB"Common Desktop Environment"\fR.
.RE

.sp
.ne 2
.na
\fB\fBSUNW_PRODVERS\fR\fR
.ad
.sp .6
.RS 4n
Solaris-only parameter indicating the version or release of the product
described in \fBSUNW_PRODNAME\fR (maximum length of 256 \fBASCII\fR
characters). For example, where \fBSUNW_PRODNAME="\fR\fISunOS\fR\fB"\fR, and
the Solaris 2.x Beta release, this string could be \fB"5.x BETA"\fR, while for
the Solaris 2.x FCS release, the string would be \fB"5.x"\fR. For Solaris 10,
the string is \fB"5.10"\fR. If the \fBSUNW_PRODNAME\fR parameter is \fINULL\fR,
so should be the \fBSUNW_PRODVERS\fR parameter.
.RE

.sp
.ne 2
.na
\fB\fBULIMIT\fR\fR
.ad
.sp .6
.RS 4n
If set, this parameter is passed as an argument to the \fBulimit\fR(1) command
(see \fBlimit\fR(1)), which establishes the maximum size of a file during
installation.
.RE

.sp
.ne 2
.na
\fB\fBVENDOR\fR\fR
.ad
.sp .6
.RS 4n
Used to identify the vendor that holds the software copyright (maximum length
of 256 \fBASCII\fR characters).
.RE

.sp
.ne 2
.na
\fB\fBVSTOCK\fR\fR
.ad
.sp .6
.RS 4n
The vendor stock number, if any, that identifies this product (maximum length
of 256 \fBASCII\fR characters).
.RE

.sp
.LP

usr/src/lib/libadm/common/pkginfo.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/


/*
 * Copyright 2005 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*LINTLIBRARY*/

/*  5-20-92   added newroot functions  */

#include <stdio.h>
#include <limits.h>
#include <stdarg.h>
#include <unistd.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <pkginfo.h>
#include <pkgstrct.h>
#include <pkglocs.h>
#include <errno.h>
#include "libadm.h"

static void	initpkg(struct pkginfo *);
static int	rdconfig(struct pkginfo *, char *, char *);
static int	ckinfo(char *, char *, char *);
static int	ckinst(char *, char *, char *, char *, char *);
static int	verscmp(char *, char *);
static int	archcmp(char *, char *);
static int	compver(char *, char *);

/*
 * Globals:
 *	pkgdir - specifies the directory where information about packages
 *	    resides, i.e. the pkginfo file is located in a subdirectory
 *
 * Caveats:
 *	The structure provided via "info" will contain malloc'd information;
 *	    this will be free'd upon the next call to pkginfo with this
 *	    same structure.  Application calls must make sure this structure
 *	    is null on the first call, or else we'll free static memory areas
 *	If the "pkg" argument is a wildcard specification, the next found
 *	    instance available which matches the request will be returned
 *	If the "pkg" argument is a NULL pointer, the structure pointed to
 *	    via "info" will have its elements deallocated and all files
 *	    associated with this routine will be closed
 *
 * Return codes:
 *	A non-zero exit code indicates error with "errno" appropriately set:
 *	    EINVAL - invalid argument
 *	    ESRCH - there are no more instances of this package around
 *	    EACCESS - unable to access files which should have been there
 */

/*VARARGS*/
int
pkginfo(struct pkginfo *info, char *pkginst, ...)
{
	char	*ckarch, *ckvers;
	int	check;
	va_list ap;

	va_start(ap, pkginst);
	if (info == NULL) {
		errno = EINVAL;
		return (-1);
	}
	if (pkginst == NULL) {
		info->pkginst = NULL;
		(void) fpkginfo(info, NULL);
		(void) fpkginst(NULL);
		return (0);
	}
	ckarch = va_arg(ap, char *);
	ckvers = va_arg(ap, char *);
	va_end(ap);

	check = 0;
	if (pkgnmchk(pkginst, "all", 1)) {
		/* wild card specification */
		pkginst = fpkginst(pkginst, ckarch, ckvers);
		if (pkginst == NULL)
			return (-1);
	} else {
		/* request to check indicated instance */
		if (ckarch || ckvers)
			check++;
	}

	info->pkginst = NULL;
	if (fpkginfo(info, pkginst))
		return (-1);

	if (check) {
		/*
		 * verify that the provided instance matches
		 * any arch & vers specs that were provided
		 */
		if (ckinst(pkginst, info->arch, info->version, ckarch,
		    ckvers)) {
			errno = ESRCH;
			return (-1);
		}
	}
	return (0);
}
/*ARGSUSED*/

int
fpkginfo(struct pkginfo *info, char *pkginst)
{

	if (info == NULL) {
		errno = EINVAL;
		return (-1);
	}

	initpkg(info);

	if (pkginst == NULL)
		return (0);
	else if (pkgnmchk(pkginst, "all", 1)) {
		errno = EINVAL; /* not an instance identifier */
		return (-1);
	}
	if (pkgdir == NULL)
		pkgdir = get_PKGLOC();

	if (rdconfig(info, pkginst, NULL)) {
		initpkg(info);
		return (-1);
	}
	return (0);
}

static void
initpkg(struct pkginfo *info)
{
	/* free previously allocated space */
	if (info->pkginst) {
		free(info->pkginst);
		free(info->arch);
		free(info->version);
		free(info->basedir);
		free(info->name);
		free(info->vendor);
		free(info->catg);
	}

	info->pkginst = NULL;
	info->arch = info->version = NULL;
	info->basedir = info->name = NULL;
	info->vendor = info->catg = NULL;
	info->status = PI_UNKNOWN;
}

static int
rdconfig(struct pkginfo *info, char *pkginst, char *ckvers)
{
	FILE	*fp;
	char	temp[256];
	char	*value, *pt, *copy, **memloc;
	int	count;

	if ((fp = pkginfopen(pkgdir, pkginst)) == NULL) {
		errno = EACCES;
		return (-1);
	}

	*temp = '\0';
	count = 0;
	while (value = fpkgparam(fp, temp)) {
		if (strcmp(temp, "ARCH") == 0 ||
		    strcmp(temp, "CATEGORY") == 0) {
			/* remove all whitespace from value */
			pt = copy = value;
			while (*pt) {
				if (!isspace((unsigned char)*pt))
					*copy++ = *pt;
				pt++;
			}
			*copy = '\0';
		}
		count++;
		memloc = NULL;
		if (strcmp(temp, "NAME") == 0)
			memloc = &info->name;
		else if (strcmp(temp, "VERSION") == 0)
			memloc = &info->version;
		else if (strcmp(temp, "ARCH") == 0)
			memloc = &info->arch;
		else if (strcmp(temp, "VENDOR") == 0)
			memloc = &info->vendor;
		else if (strcmp(temp, "BASEDIR") == 0)
			memloc = &info->basedir;
		else if (strcmp(temp, "CATEGORY") == 0)
			memloc = &info->catg;

		temp[0] = '\0';
		if (memloc == NULL)
			continue; /* not a parameter we're looking for */

		*memloc = strdup(value);
		if (!*memloc) {
			(void) fclose(fp);
			errno = ENOMEM;
			return (-1); /* malloc from strdup failed */
		}
	}
	(void) fclose(fp);

	if (!count) {
		errno = ESRCH;
		return (-1);
	}

	info->status = (strcmp(pkgdir, get_PKGLOC()) ? PI_SPOOLED :
	    PI_INSTALLED);

	if (info->status == PI_INSTALLED) {
		(void) snprintf(temp, sizeof (temp),
		    "%s/%s/!I-Lock!", pkgdir, pkginst);
		if (access(temp, 0) == 0)
			info->status = PI_PARTIAL;
		else {
			(void) snprintf(temp, sizeof (temp),
			    "%s/%s/!R-Lock!", pkgdir, pkginst);
			if (access(temp, 0) == 0)
				info->status = PI_PARTIAL;
		}
	}
	info->pkginst = strdup(pkginst);
	return (0);
}

static int
ckinst(char *pkginst, char *pkgarch, char *pkgvers, char *ckarch, char *ckvers)
{
	if (ckarch && archcmp(ckarch, pkgarch))
		return (-1);
	if (ckvers) {
		/* Check for exact version match */
		if (verscmp(ckvers, pkgvers)) {
			/* Check for compatable version */
			if (compver(pkginst, ckvers))
				return (-1);
		}
	}
	return (0);
}

/*VARARGS*/
char *
fpkginst(char *pkg, ...)
{
	static char pkginst[PKGSIZ+1];
	static DIR *pdirfp;
	struct dirent64 *dp;
	char	*ckarch, *ckvers;
	va_list	ap;

	va_start(ap, pkg);

	if (pkg == NULL) {
		/* request to close or rewind the file */
		if (pdirfp) {
			(void) closedir(pdirfp);
			pdirfp = NULL;
		}
		return (NULL);
	}

	ckarch = va_arg(ap, char *);
	ckvers = va_arg(ap, char *);
	va_end(ap);

	if (!pkgdir)
		pkgdir = get_PKGLOC();

	if (!pdirfp && ((pdirfp = opendir(pkgdir)) == NULL)) {
		errno = EACCES;
		return (NULL);
	}

	while ((dp = readdir64(pdirfp)) != NULL) {
		if (dp->d_name[0] == '.')
			continue;

		if (pkgnmchk(dp->d_name, pkg, 0))
			continue; /* ignore invalid SVR4 package names */

		if (ckinfo(dp->d_name, ckarch, ckvers))
			continue;

		/*
		 * Leave directory open in case user requests another
		 * instance.
		 */
		(void) strcpy(pkginst, dp->d_name);
		return (pkginst);
	}

	errno = ESRCH;
	/* close any file we might have open */
	(void) closedir(pdirfp);
	pdirfp = NULL;
	return (NULL);
}

static int
verscmp(char *request, char *actual)
{
	/* eat leading white space */
	while (isspace((unsigned char)*actual))
		actual++;
	while (isspace((unsigned char)*request))
		request++;

	while (*request || *actual) {
		/*
		 * Once the pointers don't match, return an error condition.
		 */

		if (*request++ != *actual++)
			return (-1);

		/* eat white space if any in both the strings */
		if (isspace((unsigned char)*request)) {
			if (*actual && !isspace((unsigned char)*actual))
				return (-1);
			while (isspace((unsigned char)*request))
				request++;
			while (isspace((unsigned char)*actual))
				actual++;
		}
	}

	return (0);

}

static int
compver(char *pkginst, char *version)
{
	FILE *fp;
	char temp[256];

	(void) snprintf(temp, sizeof (temp),
	    "%s/%s/install/compver", get_PKGLOC(), pkginst);
	if ((fp = fopen(temp, "r")) == NULL)
		return (-1);

	while (fgets(temp, 256, fp)) {
		if (*temp == '#')
			continue;
		if (verscmp(temp, version) == 0) {
			(void) fclose(fp);
			return (0);
		}
	}
	(void) fclose(fp);
	return (-1);
}

static int
archcmp(char *arch, char *archlist)
{
	char *pt;

	if (arch == NULL)
		return (0);

	/* arch and archlist must not contain whitespace! */

	while (*archlist) {
		for (pt = arch; *pt && (*pt == *archlist); )
			pt++, archlist++;
		if (!*pt && (!*archlist || (*archlist == ',')))
			return (0);
		while (*archlist) {
			if (*archlist++ == ',')
				break;
		}
	}
	return (-1);
}

static int
ckinfo(char *inst, char *arch, char *vers)
{
	FILE	*fp;
	char	temp[128];
	char	file[PATH_MAX];
	char	*pt, *copy, *value, *myarch, *myvers;
	int	errflg;

	(void) snprintf(file, sizeof (file), "%s/%s/pkginfo", pkgdir, inst);
	if ((fp = fopen(file, "r")) == NULL)
		return (1);

	if ((arch == NULL) && (vers == NULL)) {
		(void) fclose(fp);
		return (0);
	}
	temp[0] = '\0';
	myarch = myvers = NULL;
	while (value = fpkgparam(fp, temp)) {
		if (strcmp(temp, "ARCH") == 0) {
			/* remove all whitespace from value */
			pt = copy = value;
			while (*pt) {
				if (!isspace((unsigned char)*pt))
					*copy++ = *pt;
				pt++;
			}
			*copy = '\0';
			myarch = value;
			if (myvers)
				break;
		} else if (strcmp(temp, "VERSION") == 0) {
			myvers = value;
			if (myarch)
				break;
		} else
			free(value);
		temp[0] = '\0';
	}
	(void) fclose(fp);
	errflg = 0;

	if (ckinst(inst, myarch, myvers, arch, vers))
		errflg++;

	free(myarch);
	free(myvers);

	return (errflg);
}
